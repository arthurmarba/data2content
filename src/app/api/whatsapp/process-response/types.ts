// src/app/api/whatsapp/process-response/types.ts
// MODIFICADO: v1.3 - Adicionado campo opcional igConnected ao ProcessRequestBody para CTA de conexão do Instagram.
// MODIFICADO: v1.2.1 - Confirmada inclusão de instagramMediaId em PostObject para resolver erro de tipo.
// MODIFICADO: v1.2 - Tornar dialogueState opcional em EnrichedAIContext.
// MODIFICADO: v1.1 - Adicionado currentAlertDetails a EnrichedAIContext

import { DeterminedIntent } from '@/app/lib/intentService';
import { IUser, AlertDetails } from '../../../models/User';
import { IDialogueState } from '@/app/lib/stateService';

/**
 * Define a estrutura esperada para o corpo da requisição
 * processada pelo QStash worker.
 */
export interface ProcessRequestBody {
  fromPhone?: string;
  incomingText?: string;
  userId: string;
  taskType?: string;
  determinedIntent: DeterminedIntent | null;
  qstashMessageId?: string;

  /** NOVO: indica se o IG está conectado; vindo do incoming ou preenchido no worker */
  igConnected?: boolean;
}

/**
 * Representa a estrutura de um objeto de post como usado dentro
 * da lógica de detecção de alertas e outras manipulações nesta rota.
 * O campo instagramMediaId está incluído aqui.
 */
export interface PostObject {
  _id: string;
  userId: string;
  platformPostId?: string;
  instagramMediaId?: string; // Campo necessário para reportService.ts
  type: 'IMAGE' | 'CAROUSEL' | 'REEL' | 'VIDEO' | 'STORY' | string;
  description?: string;
  createdAt: Date | string;
  postDate?: Date | string;
  totalImpressions?: number;
  totalEngagement?: number;
  videoViews?: number;
  totalComments?: number;

  format?: string;
  proposal?: string;
  context?: string;

  tags?: string[];
}

/**
 * Estrutura para um evento detectado pelo Radar Mobi,
 * a ser usado para gerar a mensagem para a IA e para logs.
 */
export interface DetectedEvent {
  type: string;
  messageForAI: string;
  detailsForLog: AlertDetails;
}

/**
 * Contexto enriquecido passado para as funções da LLM.
 * MODIFICADO: dialogueState agora é opcional.
 */
export interface EnrichedAIContext {
  user: IUser;
  historyMessages: any[];
  dialogueState?: IDialogueState;
  userName: string;
  currentAlertDetails?: AlertDetails;
}
