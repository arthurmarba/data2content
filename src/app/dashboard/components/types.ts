// src/app/dashboard/components/types.ts

export interface Filtros {
    periodos?: string[];       // Ex: ["7", "30", "60", "200"]
    conteudo?: string;         // Ex: "reels até 15s"
    // Adicione outros filtros conforme necessário
  }
  
  export interface DashboardPayload {
    visao: string;
    missao: string;
    objetivos: string[];       // Lista de tags ou objetivos
    filtros?: Filtros;
    // Mantemos também os campos antigos, se necessário:
    metrics?: Record<string, number | string>;
    question?: string;
  }
  
  // Interface para o retorno da IA (customData)
  export interface CustomData {
    metrics: Record<string, number | string>;
    recommendations: Record<string, string>;
    summary: string;
    indicators: Indicator[];
  }
  
  export interface Indicator {
    id: string;                // Identificador único para o indicador
    title: string;
    value: number | string;
    description: string;
    // Você pode incluir outras propriedades para configuração gráfica se precisar
  }
  