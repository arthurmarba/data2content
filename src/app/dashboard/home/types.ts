// src/app/dashboard/home/types.ts
// Tipos compartilhados pelos cards da Home.

export interface NextPostCardData {
  slotLabel: string;
  primaryHook: string;
  secondaryHooks?: string[];
  expectedLiftPercent?: number | null;
  plannerUrl?: string;
  plannerSlotId?: string | null;
  slotDateIso?: string | null;
  isInstagramConnected: boolean;
}

export interface ConsistencyCardData {
  streakDays: number;
  weeklyGoal: number;
  postsSoFar: number;
  projectedPosts?: number;
  overpostingWarning?: boolean;
  plannerUrl?: string;
  hotSlotsUrl?: string;
}

export interface MentorshipCardData {
  nextSessionLabel: string;
  topic?: string;
  description?: string;
  joinCommunityUrl?: string;
  calendarUrl?: string;
  whatsappReminderUrl?: string;
  isMember: boolean;
}

export interface MediaKitCardData {
  shareUrl?: string;
  highlights: Array<{
    label: string;
    value: string;
  }>;
  lastUpdatedLabel?: string;
  hasMediaKit: boolean;
}

export interface CommunityMetricItem {
  id: string;
  label: string;
  value: string;
  deltaPercent?: number | null;
  periodLabel: string;
}

export interface CommunityMetricsCardData {
  metrics: CommunityMetricItem[];
  period: "7d" | "30d" | "90d";
}

export interface HomePlanSummary {
  status: string | null;
  normalizedStatus: string;
  interval: "month" | "year" | null;
  cancelAtPeriodEnd: boolean;
  expiresAt?: string | null;
  priceId?: string | null;
  hasPremiumAccess: boolean;
  isPro: boolean;
  trial: {
    active: boolean;
    eligible: boolean;
    started: boolean;
    expiresAt?: string | null;
  };
}

export interface HomeWhatsAppSummary {
  linked: boolean;
  phone?: string | null;
  startUrl?: string | null;
  trial: {
    active: boolean;
    eligible: boolean;
    started: boolean;
    expiresAt?: string | null;
  };
}

export interface HomeCommunitySummary {
  free: {
    isMember: boolean;
    inviteUrl?: string | null;
  };
  vip: {
    hasAccess: boolean;
    isMember: boolean;
    inviteUrl?: string | null;
  };
}

export interface HomeGoalsSummary {
  weeklyPostsTarget?: number | null;
  currentStreak?: number | null;
}

export interface HomeSummaryResponse {
  nextPost?: NextPostCardData | null;
  consistency?: ConsistencyCardData | null;
  mentorship?: MentorshipCardData | null;
  mediaKit?: MediaKitCardData | null;
  communityMetrics: CommunityMetricsCardData;
  plan?: HomePlanSummary | null;
  whatsapp?: HomeWhatsAppSummary | null;
  community?: HomeCommunitySummary | null;
  goals?: HomeGoalsSummary | null;
}
