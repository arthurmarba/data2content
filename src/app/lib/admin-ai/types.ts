/**
 * @fileoverview Tipos, interfaces e constantes compartilhadas para a IA do Admin.
 * @description Centraliza as definições de tipo para garantir consistência
 * em toda a lógica da Central de Inteligência.
 */

import type { ChatCompletionMessageParam } from 'openai/resources/chat/completions';

// --- Constantes de Serviço ---
// ATUALIZADO: Versão alinhada com o orquestrador de produção.
export const SERVICE_TAG = '[adminAiOrchestrator v13.0.0]';
export const DATA_DELIMITER = '\n---JSON_DATA_PAYLOAD---';
export const STATUS_DELIMITER = '---STATUS_UPDATE---';
export const MAX_TOOL_ITERATIONS = 5;

// --- Interfaces de Dados ---

/**
 * Contexto da requisição, contendo informações sobre o usuário administrador.
 */
export interface AdminAIContext {
  adminName: string;
}

/**
 * Objeto de retorno da função principal do orquestrador.
 */
export interface AdminLLMResult {
  stream: ReadableStream<string>;
}

/**
 * Representa os dados estruturados extraídos da resposta da IA
 * para renderização de componentes no frontend.
 */
export interface ExtractedDataPayload {
  visualizations: any[];
  suggestions: any[];
}

/**
 * Representa o conteúdo de uma mensagem de ferramenta, que pode
 * conter tanto o sumário para a IA quanto os dados para o frontend.
 */
export interface ToolFunctionResult {
  summary: string;
  visualizations?: any[];
  suggestions?: any[];
}
