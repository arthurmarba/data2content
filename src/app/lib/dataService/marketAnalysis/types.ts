/**
 * @fileoverview Tipos e interfaces compartilhados para o marketAnalysisService.
 * @version 1.1.0
 * @description Adicionados tipos para ranking de categoria genérico e removidos os tipos específicos de proposta que se tornaram obsoletos.
 */

import { Types } from 'mongoose';
import { z } from 'zod';

// --- (NOVO) Tipos para Ranking de Categoria Genérico ---
export type RankableCategory = 'proposal' | 'format' | 'context';

export interface ICategoryMetricRankItem {
  category: string;
  value: number;
}
// --- FIM DA SEÇÃO NOVA ---

// --- Validação de Schema e Tipos de Contrato ---

export const TopCreatorMetricEnum = z.enum([
    'total_interactions',
    'engagement_rate_on_reach',
    'likes',
    'shares',
    'comments'
]);
export type TopCreatorMetric = z.infer<typeof TopCreatorMetricEnum>;

// --- Allowed metrics for category ranking ---
export const CategoryRankingMetricEnum = z.enum([
  'shares',
  'likes',
  'comments',
  'reach',
  'views',
  'posts'
]);
export type CategoryRankingMetric = z.infer<typeof CategoryRankingMetricEnum>;

// --- (REMOVIDO) O Enum e o Tipo abaixo foram substituídos pela nova abordagem genérica ---
// export const ProposalRankingMetricEnum = z.enum([
//   'avg_views',
//   'total_interactions'
// ]);
// export type ProposalRankingMetric = z.infer<typeof ProposalRankingMetricEnum>;
// -----------------------------------------------------------------------------------------

// --- Interfaces de Contrato ---

export interface IMarketPerformanceResult {
    avgEngagementRate?: number;
    avgShares?: number;
    avgLikes?: number;
    postCount: number;
}

export interface ITopCreatorResult {
    creatorId: string;
    creatorName?: string;
    metricValue: number;
    totalInteractions: number;
    postCount: number;
}

export interface ICreatorProfile {
    creatorId: string;
    creatorName: string;
    postCount: number;
    avgLikes: number;
    avgShares: number;
    avgEngagementRate: number;
    topPerformingContext: string;
    profilePictureUrl?: string;
}

export interface IFetchMultipleCreatorProfilesArgs {
  creatorIds: string[];
}

export interface FindGlobalPostsArgs {
    context?: string;
    proposal?: string;
    format?: string;
    minInteractions?: number;
    limit?: number;
    page?: number;
    sortBy?: string;
    sortOrder?: 'asc' | 'desc';
    dateRange?: {
        startDate?: Date;
        endDate?: Date;
    };
}

export interface IGlobalPostResult {
  _id: Types.ObjectId;
  text_content?: string;
  description?: string;
  creatorName?: string;
  postDate?: Date;
  format?: string;
  proposal?: string;
  context?: string;
  stats?: {
    total_interactions?: number;
    likes?: number;
    shares?: number;
  };
}

export interface IGlobalPostsPaginatedResult {
    posts: IGlobalPostResult[];
    totalPosts: number;
    page: number;
    limit: number;
}

export interface IFetchDashboardCreatorsListParams {
  page?: number;
  limit?: number;
  sortBy?: string;
  sortOrder?: 'asc' | 'desc';
  filters?: {
    nameSearch?: string;
    planStatus?: string[];
    expertiseLevel?: string[];
    minTotalPosts?: number;
    minFollowers?: number;
    startDate?: string;
    endDate?: string;
  };
}

export interface IDashboardCreator {
  _id: Types.ObjectId;
  name: string;
  planStatus?: string;
  inferredExpertiseLevel?: string;
  totalPosts: number;
  lastActivityDate?: Date;
  avgEngagementRate: number;
  profilePictureUrl?: string;
  recentAlertsSummary?: {
    count: number;
    alerts: Array<{ type: string; date: Date; message?: string }>;
  };
}

export interface IFetchDashboardOverallContentStatsFilters {
  dateRange?: {
    startDate?: Date;
    endDate?: Date;
  };
}

export interface IDashboardOverallStats {
  totalPlatformPosts: number;
  averagePlatformEngagementRate: number;
  totalContentCreators: number;
  breakdownByFormat: { format: string; count: number }[];
  breakdownByProposal: { proposal: string; count: number }[];
  breakdownByContext: { context: string; count: number }[];
}


export interface IFetchTucaRadarEffectivenessArgs {
    alertType?: string;
    periodDays: number;
}

export interface ITucaRadarEffectivenessResult {
    alertType: string;
    positiveInteractionRate: number;
    totalAlerts: number;
}

export interface IFetchCohortComparisonArgs {
    metric: string;
    cohorts: { filterBy: 'planStatus' | 'inferredExpertiseLevel'; value: string }[];
    dateRange?: {
        startDate: Date;
        endDate: Date;
    };
}

export interface ICohortComparisonResult {
    cohortName: string;
    avgMetricValue: number;
    userCount: number;
}

export interface IRankingCreatorInfo {
  creatorId: Types.ObjectId;
  creatorName: string;
  profilePictureUrl?: string;
}

export interface ICreatorMetricRankItem extends IRankingCreatorInfo {
  metricValue: number;
}

// --- (REMOVIDO) Substituído por ICategoryMetricRankItem para ser genérico ---
// export interface IProposalMetricRankItem {
//   proposal: string;
//   metricValue: number;
// }
// --------------------------------------------------------------------------

export interface IFetchCreatorRankingParams {
  dateRange: {
    startDate: Date;
    endDate: Date;
  };
  limit?: number;
}

export interface IFetchCreatorTimeSeriesArgs {
  creatorId: string;
  metric: 'post_count' | 'avg_engagement_rate' | 'avg_likes' | 'avg_shares' | 'total_interactions';
  period: 'monthly' | 'weekly';
  dateRange: { startDate: Date; endDate: Date };
}

export interface ISegmentDefinition {
  format?: string;
  proposal?: string;
  context?: string;
}

export interface IFetchSegmentPerformanceArgs {
  criteria: ISegmentDefinition;
  dateRange: { startDate: Date; endDate: Date };
}

export interface ISegmentPerformanceResult {
  postCount: number;
  avgEngagementRate: number;
  avgLikes: number;
  avgShares: number;
  avgComments: number;
}

export interface IPeriod {
  startDate: Date;
  endDate: Date;
}

export type TopMoverEntityType = 'content' | 'creator';

export type TopMoverMetric =
  | 'cumulativeViews'
  | 'cumulativeLikes'
  | 'cumulativeShares'
  | 'cumulativeComments'
  | 'cumulativeSaved'
  | 'cumulativeReach'
  | 'cumulativeImpressions'
  | 'cumulativeTotalInteractions';

export type TopMoverSortBy =
  | 'absoluteChange_increase'
  | 'absoluteChange_decrease'
  | 'percentageChange_increase'
  | 'percentageChange_decrease';

export interface ITopMoverCreatorFilters {
    planStatus?: string[];
    inferredExpertiseLevel?: string[];
}

export interface IFetchTopMoversArgs {
  entityType: TopMoverEntityType;
  metric: TopMoverMetric;
  currentPeriod: IPeriod;
  previousPeriod: IPeriod;
  topN?: number;
  sortBy?: TopMoverSortBy;
  creatorFilters?: ITopMoverCreatorFilters;
  contentFilters?: ISegmentDefinition;
}

export interface ITopMoverResult {
  entityId: string;
  entityName: string;
  profilePictureUrl?: string;
  coverUrl?: string; // <<< ADICIONE ESTA LINHA
  metricName: TopMoverMetric;
  previousValue: number;
  currentValue: number;
  absoluteChange: number;
  percentageChange: number | null;
}

export interface IFetchPlatformConversionMetricsArgs {
  dateRange: {
    startDate: Date;
    endDate: Date;
  };
}

export interface ICreatorTimeSeriesDataPoint {
  date: Date;
  value: number;
}

export interface IPlatformConversionMetrics {
  averageFollowerConversionRatePerPost: number;
  accountFollowerConversionRate: number;
  numberOfPostsConsideredForRate: number;
  accountsEngagedInPeriod: number;
  followersGainedInPeriod: number;
}