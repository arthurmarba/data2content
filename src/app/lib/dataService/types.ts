/**
 * @fileoverview Definições de tipos e interfaces para o dataService.
 * @version 2.14.14 (Adiciona instagramMediaId a PostObject)
 * @version 2.14.13 (Atualiza CommunityInspirationFilters e outros para usar Enums)
 */
import { Types } from 'mongoose';

// Importar tipos de helpers do seu projeto
import type {
    OverallStats,
    DurationStat,
    DetailedContentStat,
    ProposalStat,
    ContextStat,
    PerformanceByDayPCO,
    DayOfWeekStat
} from '@/app/lib/reportHelpers';

// Importar IMetric e IMetricStats localmente para uso interno
import type { IMetric as LocalIMetric, IMetricStats as LocalIMetricStats } from '@/app/models/Metric';

// Importar os tipos Enum para a Comunidade de Inspiração
import type {
    FormatType,
    ProposalType,
    ContextType,
    QualitativeObjectiveType,
    PerformanceHighlightType
} from '@/app/lib/constants/communityInspirations.constants';

// Reexportar tipos dos modelos Mongoose
export type {
    IUser,
    UserExpertiseLevel,
    IUserPreferences,
    IUserLongTermGoal,
    IUserKeyFact,
    IAlertHistoryEntry,
    AlertDetails,
} from '@/app/models/User';

export type {
    IMetric,
    IMetricStats,
} from '@/app/models/Metric';

export type {
    IAdDeal,
} from '@/app/models/AdDeal';

export type {
    IAccountInsight,
} from '@/app/models/AccountInsight';

export type {
    ICommunityInspiration,
    IInternalMetricsSnapshot,
} from '@/app/models/CommunityInspiration';

export type {
    IDailyMetricSnapshot,
} from '@/app/models/DailyMetricSnapshot';

// --- Tipos para Dados de Crescimento Detalhados ---
export interface HistoricalGrowth {
  followerChangeShortTerm?: number;
  followerGrowthRateShortTerm?: number;
  avgEngagementPerPostShortTerm?: number;
  avgReachPerPostShortTerm?: number;
}

export interface LongTermGrowth {
  monthlyFollowerTrend?: { month: string; followers: number }[];
  monthlyReachTrend?: { month: string; avgReach: number }[];
  monthlyEngagementTrend?: { month: string; avgEngagement: number }[];
}

export interface IGrowthDataResult {
  historical: HistoricalGrowth;
  longTerm: LongTermGrowth;
  dataIsPlaceholder?: boolean;
  reasonForPlaceholder?: string;
}

// --- Tipos e Interfaces Específicos do DataService ---
export interface IEnrichedReport {
    overallStats?: OverallStats;
    profileSegment: string;
    multimediaSuggestion: string;
    top3Posts?: Pick<LocalIMetric, '_id' | 'description' | 'postLink' | 'stats' | 'type' | 'format' | 'proposal' | 'context' | 'instagramMediaId'>[]; // Adicionado instagramMediaId
    bottom3Posts?: Pick<LocalIMetric, '_id' | 'description' | 'postLink' | 'stats' | 'type' | 'format' | 'proposal' | 'context' | 'instagramMediaId'>[]; // Adicionado instagramMediaId
    durationStats?: DurationStat[];
    detailedContentStats?: DetailedContentStat[];
    proposalStats?: ProposalStat[];
    contextStats?: ContextStat[];
    historicalComparisons?: HistoricalGrowth;
    longTermComparisons?: LongTermGrowth;
    performanceByDayPCO?: PerformanceByDayPCO;
    recentPosts?: PostObject[];
    dayOfWeekStats?: DayOfWeekStat[];
}

export interface PreparedData {
    enrichedReport: IEnrichedReport;
}

export interface ReferenceSearchPostData {
    _id: string;
    description: string;
    proposal?: ProposalType;
    context?: ContextType;
    format?: FormatType;
}

export type ReferenceSearchResult =
    | { status: 'found'; post: ReferenceSearchPostData }
    | { status: 'clarify'; message: string }
    | { status: 'error'; message: string };

export interface IGrowthComparisons {
    weeklyFollowerChange?: number;
    monthlyReachTrend?: 'up' | 'down' | 'stable';
}

export interface AdDealInsights {
    period: 'last30d' | 'last90d' | 'all';
    totalDeals: number;
    totalRevenueBRL: number;
    averageDealValueBRL?: number;
    commonBrandSegments: string[];
    avgValueByCompensation?: { [key: string]: number };
    commonDeliverables: string[];
    commonPlatforms: string[];
    dealsFrequency?: number;
}

/**
 * Filtros para buscar inspirações na comunidade.
 * ATUALIZADO v2.14.13: Usa tipos Enum para campos categóricos.
 */
export interface CommunityInspirationFilters {
  proposal?: ProposalType;
  context?: ContextType;
  format?: FormatType;
  primaryObjectiveAchieved_Qualitative?: QualitativeObjectiveType;
  performanceHighlights_Qualitative_INCLUDES_ANY?: PerformanceHighlightType[];
  performanceHighlights_Qualitative_CONTAINS?: string; 
  tags_IA?: string[];
}

/**
 * Representa um objeto de post simplificado.
 * ATUALIZADO v2.14.14: Adicionado instagramMediaId.
 * ATUALIZADO v2.14.13: Usa tipos Enum para campos categóricos.
 */
export interface PostObject {
    _id: string;
    userId: string;
    platformPostId?: string; // Pode coexistir se necessário, ou ser removido se instagramMediaId for o padrão nesta camada
    instagramMediaId?: string; // <-- ADICIONADO PARA CORRIGIR O ERRO EM reportService.ts
    type: string; 
    description?: string;
    postDate: Date | string;
    totalImpressions?: number;
    totalEngagement?: number;
    videoViews?: number;
    totalComments?: number;
    format?: FormatType;
    proposal?: ProposalType;
    context?: ContextType;
    stats?: LocalIMetricStats;
    tags?: string[];
}

/**
 * Critérios para a função findMetricsByCriteria.
 * ATUALIZADO v2.14.13: Usa tipos Enum para campos categóricos.
 */
export interface FindMetricsCriteria {
    format?: FormatType;
    proposal?: ProposalType;
    context?: ContextType;
    dateRange?: {
        start?: string; 
        end?: string;   
    };
    minLikes?: number;
    minShares?: number;
}

export interface FindMetricsCriteriaArgs {
    criteria: FindMetricsCriteria;
    limit?: number;
    sortBy?: 'postDate' | 'stats.shares' | 'stats.saved' | 'stats.likes' | 'stats.reach';
    sortOrder?: 'asc' | 'desc';
}

export interface MetricsHistoryDataset {
    label: string;
    data: number[];
}

export interface MetricsHistoryEntry {
    labels: string[];
    datasets: MetricsHistoryDataset[];
}

export interface MetricsHistory {
    engagementRate: MetricsHistoryEntry;
    propagationIndex: MetricsHistoryEntry;
    likeCommentRatio: MetricsHistoryEntry;
    saveRateOnReach: MetricsHistoryEntry;
    followerConversionRate: MetricsHistoryEntry;
    retentionRate: MetricsHistoryEntry;
    engagementDeepVsReach: MetricsHistoryEntry;
    engagementFastVsReach: MetricsHistoryEntry;
    likes: MetricsHistoryEntry;
    comments: MetricsHistoryEntry;
}

export interface FollowerTrendPoint {
    date: string;
    value: number | null;
}

export interface FollowerTrendData {
    chartData: FollowerTrendPoint[];
    insightSummary?: string;
}

export interface ReachEngagementTrendPoint {
    date: string;
    reach: number | null;
    totalInteractions: number | null;
}

export interface ReachEngagementTrendData {
    chartData: ReachEngagementTrendPoint[];
    insightSummary?: string;
}

export interface FpcTrendPoint {
    date: string;
    avgInteractions: number | null;
}

export interface FpcTrendData {
    chartData: FpcTrendPoint[];
    insightSummary?: string;
}
