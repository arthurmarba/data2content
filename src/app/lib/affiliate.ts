import { cookies } from 'next/headers';

export function resolveAffiliateCode(req: Request, bodyCode?: string) {
  const url = new URL(req.url);
  const param = url.searchParams.get('ref') || url.searchParams.get('aff');
  const cookie = cookies().get('d2c_ref')?.value || null;
  const code = bodyCode?.trim() || param || cookie || null;
  const source = bodyCode?.trim() ? 'typed' : (param ? 'url' : (cookie ? 'cookie' : null));
  return { code, source } as { code: string | null, source: 'typed' | 'url' | 'cookie' | null };
}

export function setAffiliateCookie(code: string, days = Number(process.env.AFFILIATE_ATTRIBUTION_WINDOW_DAYS || 90)) {
  const maxAge = days * 24 * 60 * 60; // seconds
  cookies().set('d2c_ref', code, { maxAge, httpOnly: false, sameSite: 'lax' });
}

export function centsMulPct(cents: number, pct: number) {
  return Math.round(cents * (pct / 100));
}
