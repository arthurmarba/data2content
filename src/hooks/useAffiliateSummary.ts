import useSWR from 'swr';
import { AffiliateSummary, AffiliateStatus } from '@/types/affiliate';

const fetcher = (url: string) =>
  fetch(url, { cache: 'no-store' }).then((r) => {
    if (!r.ok) throw new Error('Failed to fetch summary');
    return r.json();
  });

export function useAffiliateSummary() {
  const { data, error, isLoading, mutate } = useSWR<AffiliateSummary>(
    '/api/affiliate/summary',
    fetcher,
    { revalidateOnFocus: false }
  );
  return { summary: data, loading: isLoading, error, refresh: () => mutate() };
}

export function canRedeem(
  status: AffiliateStatus | undefined,
  summary: AffiliateSummary | undefined,
  cur: string,
) {
  const curNorm = cur.toUpperCase();
  const statusCur = status?.defaultCurrency?.toUpperCase();
  if (!status?.payoutsEnabled) return false;
  if (!summary) return false;
  if (statusCur && curNorm !== statusCur) return false;
  const curSummary = summary.byCurrency?.[curNorm];
  if (!curSummary) return false;
  const min = curSummary.minRedeemCents ?? 0;
  return curSummary.availableCents >= min && curSummary.debtCents === 0;
}

export type RedeemBlockReason =
  | 'needsOnboarding'
  | 'payouts_disabled'
  | 'currency_mismatch'
  | 'below_min'
  | 'has_debt';

export function getRedeemBlockReason(
  status: AffiliateStatus | undefined,
  summary: AffiliateSummary | undefined,
  cur: string,
): RedeemBlockReason | null {
  const curNorm = cur.toUpperCase();
  const statusCur = status?.defaultCurrency?.toUpperCase();
  if (!status) return null;
  if (!status.payoutsEnabled) {
    return status.needsOnboarding ? 'needsOnboarding' : 'payouts_disabled';
  }
  if (!summary) return null;
  if (statusCur && curNorm !== statusCur) return 'currency_mismatch';
  const curSummary = summary.byCurrency?.[curNorm];
  if (!curSummary) return null;
  if (curSummary.debtCents > 0) return 'has_debt';
  const min = curSummary.minRedeemCents ?? 0;
  if (curSummary.availableCents < min) return 'below_min';
  return null;
}
