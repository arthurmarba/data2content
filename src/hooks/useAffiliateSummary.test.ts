import { canRedeem } from './useAffiliateSummary';
import { AffiliateSummary, AffiliateStatus } from '@/types/affiliate';
import { REDEEM_BLOCK_MESSAGES } from '@/copy/affiliates';

describe('canRedeem', () => {
  const status: AffiliateStatus = { payoutsEnabled: true, defaultCurrency: 'BRL' } as any;
  const summary: AffiliateSummary = {
    byCurrency: {
      BRL: { availableCents: 10000, pendingCents: 0, debtCents: 0, nextMatureAt: null, minRedeemCents: 5000 },
      USD: { availableCents: 0, pendingCents: 0, debtCents: 0, nextMatureAt: null, minRedeemCents: 0 },
    },
  };

  test('allows redeem when all conditions met', () => {
    expect(canRedeem(status, summary, 'BRL')).toBe(true);
  });

  test('blocks when payouts disabled', () => {
    expect(canRedeem({ ...status, payoutsEnabled: false }, summary, 'BRL')).toBe(false);
  });

  test('blocks when debt exists', () => {
    const s: AffiliateSummary = {
      byCurrency: { ...summary.byCurrency, BRL: { ...summary.byCurrency.BRL, debtCents: 100 } },
    };
    expect(canRedeem(status, s, 'BRL')).toBe(false);
  });

  test('blocks when below minimum', () => {
    const s: AffiliateSummary = {
      byCurrency: { ...summary.byCurrency, BRL: { ...summary.byCurrency.BRL, availableCents: 1000 } },
    };
    expect(canRedeem(status, s, 'BRL')).toBe(false);
  });

  test('blocks when currency mismatch', () => {
    expect(canRedeem(status, summary, 'USD')).toBe(false);
  });
});

describe('redeem block messages', () => {
  test('snapshot of messages', () => {
    expect(REDEEM_BLOCK_MESSAGES).toMatchSnapshot();
  });
});
