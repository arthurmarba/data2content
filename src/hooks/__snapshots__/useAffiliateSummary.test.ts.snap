// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redeem block messages snapshot of messages 1`] = `
{
  "below_min": [Function],
  "currency_mismatch": [Function],
  "has_debt": [Function],
  "needsOnboarding": "Conecte sua conta Stripe para sacar.",
  "payouts_disabled": "Sua conta Stripe precisa de verificação de documentos.",
}
`;
